### PR TITLE
Add jax to numpy, pytorch and tensorflow

### PR DESCRIPTION
**Pull request by justinDeu**
# Description

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

Fixes # (issue)

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

### Screenshots
Please attach before and after screenshots of the change if applicable.

# Checklist:

- [ ] I have run the [`pre-commit` checks](https://pre-commit.com/) with `pre-commit run --all-files` (see `CONTRIBUTING.md` instructions to set it up)
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by justinDeu** [comment]
Summary of changes and still outstanding updates:
1. everything moved over to `gymnasium`
2. seems that `GenericTestEnv` changed and doesn't work like you originally set it up in `test_tf_wrapper`, remaining tests still need to be updated to use this
3. didn't make changes to `reset()`, think you meant to include `= None` for both parameters?
4. tests still fail for ints and floats because a conversion function doesn't exist for those. I don't remember how we said we would handle those cases

File: gymnasium/dev_wrappers/to_numpy.py [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+            A tuple containing the next observation, reward, done, and extra info.
+        """
+        jax_action = numpy_to_jax(action)
+        obs, reward, done, info = self.env.step(jax_action)
+
+        obs = jax_to_numpy(obs)
+
+        return obs, reward, done, info
+
+    def reset(self, **kwargs) -> Union[np.ndarray, Tuple[np.ndarray, Dict[str, Any]]]:
```
Comment (pseudo-rnd-thoughts) [inline gymnasium/dev_wrappers/to_numpy.py:115]: Update to the full `reset(self, seed: int, options: dict[str, Any]) -> tuple[ObsType, Dict[str, Any]]` and remove the "return_info" check

File: gymnasium/dev_wrappers/to_tf.py [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+        """Performs the given action in the environment.
+
+        Args:
+            action: the action to perform as a TensorFlow Tensor
+
+        Returns:
+            A tuple containing the next observation, reward, done, and extra info
+        """
+        jax_action = tf_to_jax(action)
+        obs, reward, done, info = self.env.step(jax_action)
```
Comment (pseudo-rnd-thoughts) [inline gymnasium/dev_wrappers/to_tf.py:115]: Update in the same way as `JaxToNumpy`

File: gymnasium/dev_wrappers/to_tf.py [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+
+        obs = jax_to_tf(obs)
+
+        return obs, reward, done, info
+
+    def reset(self, **kwargs) -> Union[tf.Tensor, Tuple[tf.Tensor, dict]]:
```
Comment (pseudo-rnd-thoughts) [inline gymnasium/dev_wrappers/to_tf.py:121]: Update in the same way as `JaxToNumpy`

File: gymnasium/dev_wrappers/to_torch.py [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+    """Wraps an environment so that it can be interacted with through PyTorch Tensors.
+
+    Actions must be provided as PyTorch Tensors and observations will be returned as PyTorch Tensors.
+
+    Note:
+        Extensive testing has not been done for handling the device that the Tensor is stored on as
+        well as managing the tensor's data type.
+    """
+
+    def __init__(self, env: Env, device: Optional[torch.device] = None):
```
Comment (pseudo-rnd-thoughts) [inline gymnasium/dev_wrappers/to_torch.py:107]: Should the type not be `Optional[Device]` using the definition to provide at the beginning

File: gymnasium/dev_wrappers/to_torch.py [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+        """Performs the given action within the environment.
+
+        Args:
+            action: The action to perform as a PyTorch Tensor
+
+        Returns:
+            The next observation, reward, done and extra info
+        """
+        jax_action = torch_to_jax(action)
+        obs, reward, done, info = self.env.step(jax_action)
```
Comment (pseudo-rnd-thoughts) [inline gymnasium/dev_wrappers/to_torch.py:129]: Update like `JaxToNumpy`

File: gymnasium/dev_wrappers/to_torch.py [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+
+        obs = jax_to_torch(obs, device=self.device)
+
+        return obs, reward, done, info
+
+    def reset(self, **kwargs) -> Union[torch.Tensor, Tuple[torch.Tensor, Dict]]:
```
Comment (pseudo-rnd-thoughts) [inline gymnasium/dev_wrappers/to_torch.py:135]: Update like `JaxToNumpy`

File: tests/dev_wrappers/test_np_wrapper.py [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+        self.observation_space = gym.spaces.Box(
+            low=np.array([-1.0]), high=np.array([1.0]), dtype=np.float64
+        )
+        self.returned_rewards = [0, 1, 2, 3, 4]
+        self.return_reward_idx = return_reward_idx
+        self.t = self.return_reward_idx
+
+    def step(self, action):
+        self.t += 1
+        return jnp.array([self.t]), self.t, self.t == len(self.returned_rewards), {}
```
Comment (pseudo-rnd-thoughts) [inline tests/dev_wrappers/test_np_wrapper.py:26]: Add `False` for truncation

File: tests/dev_wrappers/test_np_wrapper.py [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+
+    def reset(
```
Comment (pseudo-rnd-thoughts) [inline tests/dev_wrappers/test_np_wrapper.py:28]: Update for new definition

File: tests/dev_wrappers/test_np_wrapper.py [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+    envs = gym.vector.SyncVectorEnv(env_fns)
+    envs.reset()
+
+    (obs, *_) = envs.step(envs.action_space.sample())
+    assert all(isinstance(o, np.ndarray) for o in obs)
+
+
+@pytest.mark.parametrize(
+    "value",
+    [
```
Comment (pseudo-rnd-thoughts) [inline tests/dev_wrappers/test_np_wrapper.py:92]: Add numpy types as well as python types

File: tests/dev_wrappers/test_torch_wrapper.py [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+        self.observation_space = gym.spaces.Box(
+            low=np.array([-1.0]), high=np.array([1.0]), dtype=np.float64
+        )
+        self.returned_rewards = [0, 1, 2, 3, 4]
+        self.return_reward_idx = return_reward_idx
+        self.t = self.return_reward_idx
+
+    def step(self, action):
+        self.t += 1
+        return jnp.array([self.t]), self.t, self.t == len(self.returned_rewards), {}
```
Comment (pseudo-rnd-thoughts) [inline tests/dev_wrappers/test_torch_wrapper.py:27]: Add `False` for truncation

File: tests/dev_wrappers/test_torch_wrapper.py [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+
+    def reset(
```
Comment (pseudo-rnd-thoughts) [inline tests/dev_wrappers/test_torch_wrapper.py:29]: Update for new type definition

File: tests/dev_wrappers/test_torch_wrapper.py [hunk 12]
```diff
@@ -0,0 +1,121 @@
...
+    envs.reset()
+
+    (obs, *_) = envs.step(torch.Tensor(envs.action_space.sample()))
+    assert all(isinstance(o, torch.Tensor) for o in obs)
+
+
+@pytest.mark.parametrize(
+    "value",
+    [
+        1.0,
```
Comment (pseudo-rnd-thoughts) [inline tests/dev_wrappers/test_torch_wrapper.py:103]: Should either not be pytorch tensors

File: tests/dev_wrappers/test_torch_wrapper.py [hunk 13]
```diff
@@ -0,0 +1,121 @@
...
+            "b": 7,
+        },
+    ],
+)
+def test_torch_conversion(value):
+    assert torch_to_jax(jax_to_torch(value)) == value
+
+
+@pytest.mark.parametrize(
+    "value", [np.array([1.0, 2.0, 3.0]), jnp.array([1.0, 2.0, 3.0])]
```
Comment (pseudo-rnd-thoughts) [inline tests/dev_wrappers/test_torch_wrapper.py:118]: Why is the first array, `np.array` and the second `jnp`?

File: tests/dev_wrappers/utils.py [hunk 14]
```diff
@@ -0,0 +1,11 @@
+def is_same_types(value, expected_type):
```
Comment (pseudo-rnd-thoughts) [inline tests/dev_wrappers/utils.py:1]: Add docstring and return type

**Review by pseudo-rnd-thoughts — CHANGES_REQUESTED** [review]
Considering the problem of zero length arrays and python/numpy values, I think we should just always go for the value. 
This is as zero length arrays shouldn't really exist, we raise a warning for environments that use them and we could raise an error if we try to convert them. 
Therefore, could we add a comment in the documentation.
